README.md

```diff
@@ -1,89 +1,242 @@
+
+The diagram illustrates a distributed approach to constructing the `AggregatedMerkleSumTree`, where an orchestrator delegates tasks to executors, which are then processed in parallel by workers. The following sections will explain the roles of the Orchestrator, Executor, Worker and ExecutorSpawner.
+
+![diagram](/Orchestrator-diagram.png)
+
 ## Orchestrator
 
-WIP
+The Orchestrator in the Summa Aggregation serves as the central management component, coordinating the data processing activities. It plays a pivotal role in coordinating the activities of Executors and Workers, improving efficiency of tasks of building Merkle sum tree.
+The final result of the Orchestrator is the construction of the `AggregationMerkleSumTree`. This is achieved by aggregating the mini-trees constructed by the Workers. The Worker in here refers to a container running `mini-tree-server`.
+
+Key functions of the Orchestrator include:
+
+- **Dynamic Executor Spawning**: The Orchestrator dynamically spawns Executors in numbers set by the user. Each Executor is then connected to a dedicated Worker for efficient task execution.
+
+- **Task Management and Distribution**: It oversees the overall task flow, loading tasks and distributing them to Executors.
+
+- **Error Management and Pipeline Control**: The Orchestrator handles basic pipeline control and responds to errors by initiating the cancellation of all tasks.
+
+- **Build AggregationMerkleSumTree**: Its final result is that evaluate `AggregationMerkleSumTree` by aggregating the mini-trees generated by the Workers.
+
+## Executor and Worker
+
+The Executor acts as a crucial intermediary between the Orchestrator and Workers, facilitating the data processing workflow. Spawned by the Orchestrator, each Executor operates in a one-to-one relationship with a Worker. Its primary function is to generate a segment of the AggregationMerkleSumTree, known as a `mini-tree`, by processing entry data. These mini-trees are then aggregated by the Orchestrator to form the complete AggregationMerkleSumTree.
+
+Key aspects of the Executor's role include:
+
+- **Spawning and Connection**: Executors are dynamically spawned by the Orchestrator as part of the system's scalability. Each Executor is designed to connect with a Worker for task execution.
+
+- **Data Handling and Task Distribution**: A primary function of the Executor is to receive data entries, often parsed and prepared by the Orchestrator. Upon receiving these entries, the Executor is responsible for forwarding them to its connected Worker.
+
+- **Communication Bridge**: The Executor serves as a communication bridge within the data pipeline. It relays processed data, `mini-tree`, from Workers back to the Orchestrator.
+
+## ExecutorSpawner
+
+The `ExecutorSpawner` is responsible for initializing and terminating Executors. It can serve as the management point for creating `Executor` instances and Workers. The latter function as the `mini-tree-server`.
```

Comment:
Are workers created by the ExecutorSpawner or by their corresponding Executors?

Reply:
Also it's not obvious in the diagram - Workers seem to be created by the executors.

Reply:
> Are workers created by the ExecutorSpawner or by their corresponding Executors?

By ExecutorSpawner in here.

Reply:
> Also it's not obvious in the diagram - Workers seem to be created by the executors.

Yes, it is intentional. This is because an executor can be created using the specified node URL of an already existing worker.

Reply:
> I guess it's the ExecutorSpawner, but it wasn't clear for this description. Could you please mention that explicitly?

Okay, I think it would be better to delete 'Worker' here. Then I will explain it elsewhere.